pkg/controller/deployment/deployment_controller.go

```diff
@@ -799,28 +811,97 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
 	}
 	// Check if we can scale down.
 	minAvailable := deployment.Spec.Replicas - maxUnavailable
+	maxScaledDown := allPodsCount - minAvailable
```

Comment:
This includes both maxUnavailable and any surge pods that have been created, which is good. However, it doesn't take into account how many pods are unavailable from the newRC. 

I think we want:
```
maxScaledDown := max(allPodsCount - minAvailable - newRCPodsUnavailable, 0)
```

Concrete example:
* 10 replicas
* 2 maxUnavailable (absolute number, not percent)
* 3 maxSurge (absolute number, not percent)

Deployment is updated, newRC is created with 3 replicas, oldRC is scaled down to 8, and newRC is scaled up to 5. 

The new RC pods crashloop and never become available.

allPodsCount is 13. minAvailable is 8. newRCPodsUnavailable is 5.

A node fails and causes one of the oldRC pods to become unavailable. However, 13 - 8 - 5 = 0, so the oldRC won't be scaled down.

The user notices the crashloop and does `kubectl rollout undo` to rollback.

newRCPodsUnavailable is 1, since we rolled back to the good RC, so maxScaledDown = 13 - 8 - 1 = 4. 4 of the crashlooping pods will be scaled down. 

The total number of pods will then be 9 and the newRC can be scaled up to 10.

WDYT?